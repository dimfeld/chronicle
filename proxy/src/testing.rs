use std::{
    sync::{atomic::AtomicUsize, Arc},
    time::Duration,
};

use error_stack::{Report, ResultExt};

use crate::{
    format::{ChatChoice, ChatMessage, ChatResponse, UsageResponse},
    providers::{
        ChatModelProvider, ProviderError, ProviderErrorKind, SendRequestOptions,
        SynchronousProviderResponse,
    },
    Error,
};

#[derive(Debug)]
pub enum TestFailure {
    Transient,
    Timeout,
    BadRequest,
    RateLimit,
    Auth,
    TransformingRequest,
    TransformingResponse,
}

#[derive(Debug)]
pub struct TestProvider {
    pub name: String,
    /// Fail requests
    pub fail: Option<TestFailure>,
    pub response: String,
    pub fail_times: usize,
    pub calls: AtomicUsize,
}

impl Default for TestProvider {
    fn default() -> Self {
        Self {
            name: "test".to_string(),
            fail: None,
            response: "A response".to_string(),
            fail_times: usize::MAX,
            calls: AtomicUsize::new(0),
        }
    }
}

impl Into<Arc<dyn ChatModelProvider>> for TestProvider {
    fn into(self) -> Arc<dyn ChatModelProvider> {
        Arc::new(self)
    }
}

#[async_trait::async_trait]
impl ChatModelProvider for TestProvider {
    fn name(&self) -> &str {
        "test"
    }

    fn label(&self) -> &str {
        "Test"
    }

    async fn send_request(
        &self,
        options: SendRequestOptions,
    ) -> Result<SynchronousProviderResponse, Report<Error>> {
        let current_call = self
            .calls
            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);

        if current_call < self.fail_times {
            match self.fail {
                Some(TestFailure::Transient) => Err(ProviderErrorKind::Server),
                Some(TestFailure::Timeout) => Err(ProviderErrorKind::Timeout),
                Some(TestFailure::BadRequest) => Err(ProviderErrorKind::BadInput),
                Some(TestFailure::RateLimit) => {
                    Err(ProviderErrorKind::RateLimit { retry_after: None })
                }
                Some(TestFailure::Auth) => Err(ProviderErrorKind::AuthRejected),
                Some(TestFailure::TransformingRequest) => return Err(Error::TransformingRequest)?,
                Some(TestFailure::TransformingResponse) => {
                    return Err(Error::TransformingResponse)?
                }
                None => Ok(()),
            }
            .map_err(|kind| ProviderError {
                kind,
                status_code: None,
                body: None,
                latency: Duration::from_millis(500),
            })
            .change_context(Error::ModelError)?;
        }

        Ok(SynchronousProviderResponse {
            model: options.body.model.clone().unwrap_or_default(),
            body: ChatResponse {
                created: 1,
                model: options.body.model.clone(),
                system_fingerprint: None,
                choices: vec![ChatChoice {
                    index: 0,
                    message: ChatMessage {
                        role: Some("assistant".to_string()),
                        content: Some(self.response.clone()),
                        tool_calls: Vec::new(),
                        name: None,
                    },
                    finish_reason: "stop".to_string(),
                }],
                usage: UsageResponse {
                    prompt_tokens: Some(1),
                    completion_tokens: Some(2),
                    total_tokens: Some(3),
                },
            },
            meta: None,
            latency: Duration::from_millis(500),
        })
    }

    fn is_default_for_model(&self, _model: &str) -> bool {
        false
    }
}
