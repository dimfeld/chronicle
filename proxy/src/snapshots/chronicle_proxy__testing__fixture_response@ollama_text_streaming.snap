---
source: proxy/src/testing.rs
expression: response
---
{
  "request_info": {
    "id": "00000000-0000-0000-0000-000000000000",
    "provider": "ollama",
    "model": "me/a-test-model",
    "num_retries": 0,
    "was_rate_limited": false
  },
  "response_info": {
    "meta": {
      "eval_duration": 4726492000,
      "load_duration": 743750,
      "prompt_eval_duration": 17083000
    },
    "model": "llama3"
  },
  "was_streaming": true,
  "num_chunks": 278,
  "response": {
    "created": 0,
    "model": "llama3",
    "system_fingerprint": null,
    "choices": [
      {
        "index": 0,
        "message": {
          "role": "assistant",
          "content": "A classic!\n\nLassie is a fictional Rough Collie dog character created by Eric Knight in his 1940 novel \"Lassie Come-Home\". The story follows a collie dog named Lassie, who is a loyal companion to a young girl named Ruth Fielding. The book was a huge success, and the character of Lassie has since become an iconic symbol of courage, loyalty, and friendship.\n\nThe first television show featuring Lassie premiered in 1954 and ran for 17 seasons until 1974. The series followed the adventures of a heroic collie dog named Lassie, who lives on a ranch with her best friend Timmy Martin (played by Jon Provost) and his family. Lassie was known for her incredible intelligence, bravery, and ability to help those in need.\n\nOver the years, there have been several reboots and spin-offs of the original TV series, including a 1997-2001 revival starring Christopher Knight as Jonathan \"Jesse\" Harper, Timmy's older brother. More recently, a new Lassie TV series premiered on The Hallmark Channel in 2019.\n\nLassie has become an enduring cultural phenomenon, with numerous films, TV shows, books, and merchandise featuring the beloved character. She remains one of the most iconic and beloved dogs in American pop culture!"
        },
        "finish_reason": "stop"
      }
    ],
    "usage": {
      "prompt_tokens": null,
      "completion_tokens": 278,
      "total_tokens": null
    }
  }
}
